# Evidencia Final

### Equipo 6:

 A01706095 - Naomi Estefanía Nieto Vega

A01706189 - Alejandro Angel Calderon Berges 

A01706596 - Carlos Soria de la Cabada

### Entregable 1 - Obtención de Estadísticas Descriptivas

#### Paso 1. Cargar los datos de nuestra base de datos con ayuda de pandas

En esta sección únicamente se leen los datos, como podemos ver es acerca de unos artículos publicados en la web.

In [7]:
import pandas as pd
data = pd.read_csv("heart.csv")

In [8]:
data

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

#### Significado de las variables:

* age - edad en años <br>
* sex - (1 = hombre; 0 = mujer) <br>
* cp - tipo de dolor en el pecho previo al infarto <br>
* trestbps - presión sanguínea en mmHg <br>
* chol - colesterol en sangre en mg/dl <br>
* fbs - azucar en sangre mayor a 120 mg/dl (1 = si; 0 = no) <br>
* restecg - resultados de electrocardiograma <br>
* thalach - máxima frecuencia cardiaca <br>
* oldpeak - depresión del segmento ST <br>
* slope - pendiente del segmento ST <br>
* target - tuvo infarto o no (1=si, 0=no)

In [9]:
import pandas as pd #Aqui importamos la libreria
import csv

df=pd.read_csv("heart.csv") #Aqui leemos el archivo
df.describe() #Aqui le decimos que nos muestre los analiticos en la tabla de abajo

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

#### Paso 2. Verificar la cantidad de datos, las variables que contiene cada vector de datos e identifica el tipo de variables.

En general tenemos 302 filas con datos, de los cuales existen las variables o columas de edad, sexo, tipo de dolor de pecho, colesterol, azúcar en sangre, resultados de electrocardiograma, pulso máximo, ejercicio, depresión del segmento ST y objetivo. 

El tipo de las variables es float y entero en su mayoría

#### Paso 3. Analizar las variables para saber que representa cada una y en que rangos se encuentran. 

En este caso las variables como ya mencioné antes representan una variable de las veces que ha ocurrido un evento, en este caso relacionado con los infartos, para esto se analizan los datos relevantes como la frecuencia, el colesterol, azúcar en sangre, entre otros factores que nos permiten predecir cuando alguien podría tener un infarto basado en los datos analizados.

In [16]:
data = [['age',29.000000,77.000000,303.000000,54.366337,9.082101],
        ['sex',0.000000,1.000000,303.000000,0.683168,0.466011],
        ['cp',0.000000,3.000000,303.000000,0.966997,1.032052],
        ['trestbps',94.000000,200.000000,303.000000,0.966997,1.032052],
        ['chol',126.000000,564.124224,303.000000,246.264026,51.830751],
        ['fbs',0.000000,1.000000,303.000000,0.148515,0.356198],
        ['resteccg',0.000000,2.000000,303.000000,0.528053,0.525860],
        ['thalach',71.000000,202.000000,303.000000,149.646865,22.905161],
        ['exang',0.000000,1.000000,303.000000,0.326733,0.469794],
        ['oldpeak',0.000000,6.2000000,303.000000,1.039604,1.161075],
        ['slope',0.000000,2.000000,303.000000,1.399340,0.616226],
        ['ca',0.000000,4.000000,303.000000,0.729373,1.022606],
        ['thal',0.000000,3.000000,303.000000,2.313531,0.612277],
        ['target',0.000000,1.000000,303.000000,0.544554,0.498835]]

df = pd.DataFrame(data, columns = ['Variables', 'Min Range','Max Range','Count','Mean','std']) 
print(df)

   Variables  Min Range   Max Range  Count        Mean        std
0        age       29.0   77.000000  303.0   54.366337   9.082101
1        sex        0.0    1.000000  303.0    0.683168   0.466011
2         cp        0.0    3.000000  303.0    0.966997   1.032052
3   trestbps       94.0  200.000000  303.0    0.966997   1.032052
4       chol      126.0  564.124224  303.0  246.264026  51.830751
5        fbs        0.0    1.000000  303.0    0.148515   0.356198
6   resteccg        0.0    2.000000  303.0    0.528053   0.525860
7    thalach       71.0  202.000000  303.0  149.646865  22.905161
8      exang        0.0    1.000000  303.0    0.326733   0.469794
9    oldpeak        0.0    6.200000  303.0    1.039604   1.161075
10     slope        0.0    2.000000  303.0    1.399340   0.616226
11        ca        0.0    4.000000  303.0    0.729373   1.022606
12      thal        0.0    3.000000  303.0    2.313531   0.612277
13    target        0.0    1.000000  303.0    0.544554   0.498835


#### Paso 4. Conclusiones

Como conclusión de acuerdo a las variables podemos asumir que hay un promedio de cantidad de veces que se comparte un artículo bastante alto, siendo este de 27,948; esto podría decirnos que a la mayoría de los usuarios que interactuan les gusta ese artículo. Además es importante observar que tenemos un número alto en la desviación estándar esto quiere decir que existe una gran dispersión en la población de los datos con respecto a la media. La cuenta en general para las variables es de 161, a excepción de la cantidad de comentarios que es de 129. Asimismo tenemos una media aritmética de 1808, esto es la cantidad promedio de palabras que están escritas en los artículos.

### Entregable 2. Mapas de calor y boxplots

#### Paso 1. Cargar los datos usando tu lector de csv o con pandas. 

In [ ]:
import pandas as pd

data = pd.read_csv('heart.csv')
data.head()

#### a) Diagrama de cajas y bigotes

In [ ]:
import matplotlib.pyplot as plit
import seaborn as sns 
df.groupby('chol').sum().plot(kind='box',legend='Reverse')

En esta gráfica podemos ver el diagrama un poco disperso únicamente en 3 variables porque son las que cuentan con números más grandes o mayor variedad que se puede observar al graficarse a diferencia de las variables que tienen solo datos booleanos no hay mucho que observar. De acuerdo con esto las variables más significativas son la edad, la presión sanguínea y la frecuencia cardiaca.

#### b) Histogramas

En esta sección podemos ver el histograma, que nos muestra la frecuencia con la que ciertos hechos sucedem en este caso la frecuencia de las variables relevantes para los infartos.

In [ ]:
import matplotlib.pyplot as plt
import pandas as pd

data = pd.read_csv('heart.csv')

data.plot(kind = 'bar')
plt.ylabel('Frequencia')
plt.xlabel('Parametros')
plt.title('Infartos en personas')

plt.show()

In [ ]:
import pandas as pd
datos = pd.read_csv('heart.csv')
df = pd.DataFrame(datos)

data.drop ([0,1]).hist()

#### 3.- Mapas de calor

El mapa de color es una ayuda visual, que nos permite ver en conjunto los valores que estan asociados con colores para que sea más perceptible notar patrones. De este se eliminaron los datos con valores mínimos o booleanos ya que no mostraban un cambio relevante en la gráfica final por lo que en la línea 6 se excluyen.

In [ ]:
import pandas as pd
import seaborn as sb
import matplotlib.pyplot as plt
data = pd.read_csv('heart.csv')

uniform_data = data.drop(['sex','cp','fbs','restecg','exang','oldpeak','slope','ca','thal','target'], axis=1)
ax = sb.heatmap(uniform_data)

### Conclusiones

##### 1. ¿Hay alguna variable que no aporta información?

Si, existen variables que no se consideran muy relevantes por el tipo de dato que tienen en sus valores, esto nos impide analizarlo de esta forma y lo que debemos hacer para considerarlo es el proceso de normalización de la información para tener una base de datos limpia, sin valores extraños o valores NULL.

##### 2. Si tuvieras que eliminar variables, ¿cuáles quitarías y por qué?

Quitaría las variables de resultados de electrocardiograma, depresión y pendiente del segmento ST del ciclo cardiaco de la persona. Las borraría porque considero que los datos que aportan son muy mínimos y no tan necesarios para el alcance de análisis que tendrá nuestro proyecto en este momento.

##### 3. ¿Existen variables que tengan datos extraños?

No existen valores con caracteres extraños, unicamente los booleanos que nos complican un poco el análisis pero fuera de eso al momento de buscar la base de datos intentamos que tuviera mayormente valores númericos. Las variables con valores bool son fbs, target y el sexo.

##### 4. Si comparas las variables, ¿todas están en rangos similares? ¿Crees que esto afecte?

No todas están en rangos similares pero hay algunas que si comparten similitudes como la presión sanguínea o el colesterol en sangre ya que estos son indicios principales que nos dan una posible alerta de que esa persona puede sufrir un infarto. El hecho de que estén en rangos similares si afecta un poco ya que al estar los datos tan parecidos y no tener variedad es posible que se convierta en un patrón repetitivo de datos y esto nos limita en cuanto a predicciones. 

##### 5. ¿Puedes encontrar grupos qué se parezcan? ¿Qué grupos son estos?

Sí, como ya mencioné antes la presión sanguínea en milimetros de mercurio de las personas y el el colesterol en sangre.

### Entregable 3. Patrones con K-means

In [11]:
import pandas as pd #Aqui importamos la libreria
import csv

df=pd.read_csv("heart.csv") #Aqui leemos el archivo

countFemale = len(df[df.sex == 0])
countMale = len(df[df.sex == 1])
print("Porcentaje de pacientes mujeres: {:.2f}%".format((countFemale / (len(df.sex))*100)))
print("Porcentaje de pacientes hombres: {:.2f}%".format((countMale / (len(df.sex))*100)))

Percentage of Female Patients: 31.68%
Percentage of Male Patients: 68.32%


In [15]:
plt.scatter(x=df.age[df.target==1], y=df.thalach[(df.target==1)], c="red")
plt.scatter(x=df.age[df.target==0], y=df.thalach[(df.target==0)])
plt.legend(["Disease", "Not Disease"])
plt.xlabel("Age")
plt.ylabel("Maximum Heart Rate")
plt.show()

NameError: name 'plt' is not defined

In [14]:
# Normalize
x = (x_data - np.min(x_data)) / (np.max(x_data) - np.min(x_data)).values

x_train, x_test, y_train, y_test = train_test_split(x,y,test_size = 0.2,random_state=0)

#transpose matrices
x_train = x_train.T
y_train = y_train.T
x_test = x_test.T
y_test = y_test.T

# KNN Model
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors = 2)  # n_neighbors means k
knn.fit(x_train.T, y_train.T)
prediction = knn.predict(x_test.T)

print("{} NN Score: {:.2f}%".format(2, knn.score(x_test.T, y_test.T)*100))

NameError: name 'x_data' is not defined